backend/telemetry_shard/src/main.rs

```diff
@@ -0,0 +1,190 @@
+                let _ = tx_to_aggregator.send(FromWebsocket::Disconnected).await;
+                // Note: IF we want to close with a status code and reason, we need to construct
+                // a ws::Message using `ws::Message::close_with`, rather than using this method:
+                let _ = websocket.close().await;
+            })
+        },
+    );
+
+    // Merge the routes and start our server:
+    let routes = ws_route.or(health_route);
+    warp::serve(routes).run(opts.socket).await;
+    Ok(())
+}
+
+/// This takes care of handling messages from an established socket connection.
+async fn handle_node_websocket_connection<S>(
+    mut websocket: ws::WebSocket,
+    mut tx_to_aggregator: S,
+    addr: Option<IpAddr>,
+) -> (S, ws::WebSocket)
+where
+    S: futures::Sink<FromWebsocket, Error = anyhow::Error> + Unpin,
+{
+    // This could be a oneshot channel, but it's useful to be able to clone
+    // messages, and we can't clone oneshot channel senders.
+    let (close_connection_tx, mut close_connection_rx) = mpsc::channel(0);
+
+    // Tell the aggregator about this new connection, and give it a way to close this connection:
+    let init_msg = FromWebsocket::Initialize {
+        close_connection: close_connection_tx,
+    };
+    if let Err(e) = tx_to_aggregator.send(init_msg).await {
+        log::error!("Error sending message to aggregator: {}", e);
+        return (tx_to_aggregator, websocket);
+    }
+
+    // Now we've "initialized", wait for messages from the node. Messages will
+    // either be `SystemConnected` type messages that inform us that a new set
+    // of messages with some message ID will be sent (a node could have more
+    // than one of these), or updates linked to a specific message_id.
+    loop {
+        tokio::select! {
+            // The close channel has fired, so end the loop:
+            _ = close_connection_rx.next() => {
+                log::info!("connection to {:?} being closed by aggregator", addr);
+                break
+            },
+            // A message was received; handle it:
+            msg = websocket.next() => {
+                let msg = match msg {
+                    Some(msg) => msg,
+                    None => { log::warn!("Websocket connection from {:?} closed", addr); break }
+                };
+
+                // If we see any errors, log them and end our loop:
+                let msg = match msg {
+                    Err(e) => { log::error!("Error in node websocket connection: {}", e); break },
+                    Ok(msg) => msg,
+                };
+
+                // Close message? Break to close connection.
+                if msg.is_close() {
+                    break;
+                }
+
+                // If the message isn't something we want to handle, just ignore it.
+                // This includes system messages like "pings" and such, so don't log anything.
```

Comment:
In soketto such messages are not forwarded to consuming code I think.

Reply:
The docs suggest that pings are responded to (and I "handle" close messages above already), but since you can also check `msg.is_ping()` I thought it best to be safe and ignore any such messages if they are seen here despite being handled.